### ISSUE TITLE
dns.resolveMx might fail in specific cases in Node <22

### ISSUE THREAD

**Issue by analogic** [haraka/Haraka#3381]
Not a Haraka bug, but a pretty important node issue that affects the deliverability for Haraka outbound. So I thought it would be a good idea to report it here as well...

See this example:

```
$ docker run -ti node:22 node
...
> const dns = require('dns');
> const ign = dns.resolveMx("blek.cz", (err, addresses) => console.log([err, addresses]));

[ null, [ { exchange: 'blekmx1.blek.cz', priority: 100 } ] ]
```

```
$ docker run -ti node:20 node
...
> const dns = require('dns');
> const ign = dns.resolveMx("blek.cz", (err, addresses) => console.log([err, addresses]));

[ null, [ { exchange: 'blek.cz', priority: 100 } ] ]
```

```
$ dig mx +short blek.cz 
100 blekmx1.blek.cz.
```

Bare "blek.cz" does not have smtp open, so it bounces. I have found no irregularities in this particular DNS record - even sniffed comm looks correct.

**Comment by baudehlo** [comment]
Could this be a docker issue?

╰─ node

      ─╯

Welcome to Node.js v20.11.0.

Type ".help" for more information.

> const dns = require('dns');

undefined

> const ign = dns.resolveMx("blek.cz", (err, addresses) =>
console.log([err, addresses]));

undefined

> [ *null*, [ { exchange: 'blekmx1.blek.cz', priority: 100 } ] ]

**Comment by analogic** [comment]
Maybe, but see: 
(shortened)
```
$ docker run -ti node:16 node
[ null, [ { exchange: 'blekmx1.blek.cz', priority: 100 } ] ]

$ docker run -ti node:18 node
[ null, [ { exchange: 'blek.cz', priority: 100 } ] ]

$ docker run -ti node:20 node
[ null, [ { exchange: 'blek.cz', priority: 100 } ] ]

$ docker run -ti node:20-alpine node
[ null, [ { exchange: 'blek.cz', priority: 100 } ] ]

$ docker run -ti node:22 node
[ null, [ { exchange: 'blekmx1.blek.cz', priority: 100 } ] ]
```

It could be host system related (resolver?) and also there should be something unusual about this particular domain. I have confirmed this on Ubuntu & Red Hat

**Comment by analogic** [comment]
ref https://github.com/nodejs/node/issues/53708

**Comment by sriccio** [comment]
Hello,

I think I hit the same issue. I have delivery issues from Haraka to a specific domain:

```# dig mx +short chuv.ch
100 mail-alt.rcv.ch.
100 mail2-alt.rcv.ch.
3 mx1-pub.chuv.ch.
3 mx0-pub.chuv.ch.
```

But Haraka tries to to relay it through `chuv.ch`

```
Jul 30 19:53:04 mta-gw haraka[1483099]: [DEBUG] [FA5FA92B-389B-400C-9FFC-752823C78E75.1.1] [outbound] running get_mx hooks
Jul 30 19:53:04 mta-gw haraka[1483099]: [DEBUG] [FA5FA92B-389B-400C-9FFC-752823C78E75.1.1] [outbound] DNS (A) for chuv.ch -> 83.166.143.24
Jul 30 19:53:04 mta-gw haraka[1483099]: [DEBUG] [FA5FA92B-389B-400C-9FFC-752823C78E75.1.1] [outbound] delivering from: mta-gw-149.swisscenter.net to: 83.166.143.24:25 (0) (10)
```

It ends with "Tried all MXs" and then bounce back.

**Comment by msimerson** [comment]
It's a bug in the c-ares library used by node.js, fixed in v1.29. The practical solution for you is to upgrade your node.js to versions 22.3.0 or 20.16.0 or higher.
